sw/device/silicon_creator/lib/base/sec_mmio.c

```diff
@@ -0,0 +1,123 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/silicon_creator/lib/base/sec_mmio.h"
+
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/silicon_creator/lib/base/abs_mmio.h"
+
+// FIXME: Linker configuration.
+extern sec_mmio_ctx_t sec_mmio_ctx;
+
+// FIXME: Replace for shutdown module handler.
+static sec_mmio_shutdown_handler __shutdown_handler;
+
+// Value with good hamming weight used to mask the stored expected value.
+static const uint32_t kSecMmioMaskVal = 0x21692436u;
+
+// This must be set to a prime number greater than the number of items in
+// `sec_mmio_ctx.addrs`. Used to generate random read order permutations.
+static const uint32_t kSecMmioRndStep = 103u;
+
+/**
+ * Updates or inserts the register entry pointed to by MMIO `addr` with the
+ * given `value`.
+ *
+ * Increments the `sec_mmio_ctx.last_index`.
+ */
+static void upsert_register(uint32_t addr, uint32_t value) {
+  size_t i = 0;
+  for (; i < sec_mmio_ctx.last_index; ++i) {
+    if (sec_mmio_ctx.addrs[i] == addr) {
+      sec_mmio_ctx.values[i] = value;
+      break;
+    }
+  }
+  if (i == sec_mmio_ctx.last_index && i < kSecMmioRegFileSize) {
+    sec_mmio_ctx.addrs[i] = addr;
+    sec_mmio_ctx.values[i] = value;
+    ++sec_mmio_ctx.last_index;
+  }
+}
+
+void sec_mmio_init(sec_mmio_shutdown_handler callee) {
+  __shutdown_handler = callee;
+  sec_mmio_ctx.last_index = 0;
+  sec_mmio_ctx.write_count = 0;
+  sec_mmio_ctx.check_count = 0;
+  sec_mmio_ctx.expected_write_count = 0;
+  for (int i = 0; i < ARRAYSIZE(sec_mmio_ctx.addrs); ++i) {
+    sec_mmio_ctx.addrs[i] = UINT32_MAX;
+  }
+}
+
+uint32_t sec_mmio_read32(uint32_t addr) {
+  uint32_t value = abs_mmio_read32(addr);
+  uint32_t masked_value = value ^ kSecMmioMaskVal;
+
+  upsert_register(addr, value);
+
+  if ((abs_mmio_read32(addr) ^ kSecMmioMaskVal) != masked_value) {
+    __shutdown_handler();
+  }
+  return value;
+}
+
+void sec_mmio_write32(uint32_t addr, uint32_t value) {
+  abs_mmio_write32(addr, value);
+  uint32_t masked_value = value ^ kSecMmioMaskVal;
+
+  upsert_register(addr, masked_value);
+
+  if ((abs_mmio_read32(addr) ^ kSecMmioMaskVal) != masked_value) {
+    __shutdown_handler();
+  }
+  ++sec_mmio_ctx.write_count;
+}
+
+void sec_mmio_write_increment(uint32_t value) {
+  sec_mmio_ctx.expected_write_count += value;
+}
+
+void sec_mmio_check_values(uint32_t rnd_offset) {
+  size_t offset = rnd_offset;
+  size_t i;
+  for (i = 0; i < sec_mmio_ctx.last_index; ++i) {
+    // FIXME: Remove dependency on __udivdi3.
+    offset = (offset + kSecMmioRndStep) % sec_mmio_ctx.last_index;
+    uint32_t read_value = abs_mmio_read32(sec_mmio_ctx.addrs[offset]);
+    if ((read_value ^ kSecMmioMaskVal) != sec_mmio_ctx.values[offset]) {
+      __shutdown_handler();
+    }
+  }
+  // Check for loop completion.
+  if (i != sec_mmio_ctx.last_index) {
```

Comment:
I think it may be useful to figure out if we need a value barrier here. I suspect the function call in the `if` is sufficient, but if we were to repalce it with
```c
__shutdown_handler();
__builtin_unreachable();
```
(which we probably should) this might make the compiler think "ah yes this is only reachable if the condition was already satisfied" and delete the branch altogether.

In particular, I think it is prudent to add the following macro to hardened.h and start using it:
```c
#define OT_VALUE_BARRIER(x) asm volatile ("", "=r"(x))
```
(This is a standard trick for clobbering all compiler-deduced knowledge about `x`, because it cannot peek into the assembly block.)

Reply:
Someone internal pointed me to https://github.com/google/benchmark/blob/master/include/benchmark/benchmark.h#L328 as another phrasing of this...

Reply:
I was going to point to the Google benchmark macros too. Thanks for the boringssl link. Ultimately, both the boringssl and Google benchmark approaches were originated by Chandler Carruth who has spent quite some time thinking about this, so we'd be well advised to adopt whichever of these is most appropriate. Do note that the boringssl `value_barrier_*` is weaker than the Google benchmark `DoNotOptimize`.

There's a separate longer discussion to be had about what properties of code like this we want the compiler to maintain and how to ensure that - i.e. are the barriers like the above sufficient? Are we applying them appropriately? Do we have the necessary process in place to review that (and potentially ensure they work as expected as compilers evolve?)? etc etc. This shouldn't block landing the PR of course.

Reply:
I'm happy to start a doc on this at some point. I think it should include:
- What the barrier is for, and the extremely limited circumstances under which to use it.
- Justification on the correctness of the barrier implementation; it is unclear what optimizations we want to block, so making a list of those may be appropriate.
- Review criteria for use of the barrier.

(Obv this should not block the PR, I expect there will be a mask_rom-wide audit of this eventually.)

Reply:
Can you guys start contributing to this doc? 

https://docs.google.com/document/d/184RWslB-9baXy1c5cKDAiXcqzvwSFFGlL8XYaNoks-Y/edit

Reply:
Filed #6610 to keep track of this open.